db/view/view.cc

```diff
@@ -3597,24 +3597,58 @@ view_updating_consumer::view_updating_consumer(view_update_generator& gen, schem
     })
 { }
 
-delete_ghost_rows_visitor::delete_ghost_rows_visitor(service::storage_proxy& proxy, service::query_state& state, view_ptr view, db::timeout_clock::duration timeout_duration)
+delete_ghost_rows_visitor::delete_ghost_rows_visitor(service::storage_proxy& proxy, service::query_state& state, view_ptr view, db::timeout_clock::duration timeout_duration, size_t concurrency, promise<> done)
         : _proxy(proxy)
         , _state(state)
         , _timeout_duration(timeout_duration)
         , _view(view)
         , _view_table(_proxy.get_db().local().find_column_family(view))
         , _base_schema(_proxy.get_db().local().find_schema(_view->view_info()->base_id()))
         , _view_pk()
+        , _concurrency_semaphore(concurrency)
+        , _done(std::move(done))
 {}
 
+delete_ghost_rows_visitor::delete_ghost_rows_visitor(delete_ghost_rows_visitor&& other)
+        : _proxy(other._proxy)
+        , _state(other._state)
+        , _timeout_duration(std::move(other._timeout_duration))
+        , _view(std::move(other._view))
+        , _view_table(other._view_table)
+        , _base_schema(std::move(other._base_schema))
+        , _view_pk(std::move(other._view_pk))
+        , _concurrency_semaphore(std::move(other._concurrency_semaphore))
+        , _all_jobs(std::exchange(other._all_jobs, {}))
+        , _done(std::move(other._done))
+{}
+
+delete_ghost_rows_visitor::~delete_ghost_rows_visitor() noexcept {
+    if (_all_jobs) {
+        try {
+            _all_jobs->get();
+            _done.set_value();
+        } catch (...) {
+            _done.set_exception(std::current_exception());
+        }
+    }
+}
+
 void delete_ghost_rows_visitor::accept_new_partition(const partition_key& key, uint32_t row_count) {
     SCYLLA_ASSERT(thread::running_in_thread());
     _view_pk = key;
 }
 
 // Assumes running in seastar::thread
 void delete_ghost_rows_visitor::accept_new_row(const clustering_key& ck, const query::result_row_view& static_row, const query::result_row_view& row) {
-    auto view_exploded_pk = _view_pk->explode();
+    auto units = get_units(_concurrency_semaphore, 1).get();
+    _all_jobs = when_all_succeed(
+        std::move(_all_jobs).value(),
```

Comment:
I'm almost certain there is a better mechanism for this, but can't remember it at this point on a Friday.